docs/proposals/image-index.md

```diff
@@ -0,0 +1,153 @@
+from different parts of the system in non-unified way.
+
+This results in the tools such as pruning or `oadm top images` always build
+in-memory state of the cluster for their operation.
+
+### Goal
+
+To build efficient *Image Index API* that allows to querying the relations
+between images and also name and map these relations into human-readable form.
+
+### Known Limitations / Scope Of Work
+
+* We can only keep track of Images we know about (iow. imported into OpenShift)
+  * If somebody use an external image, we can’t map it to indexing API without importing
+    it into an Image object
+* Currently, there is no way to restrict users referencing an external images in Pods.
+  * Solving this problem goes beyond the scope of this proposal
+* This API is meant to only query the images/imagestreams not their usage in the Pods.
+* The classification is meant to be managed and used by cluster/registry administrators,
+  not by regular users
+
+## Proposed Workflow
+
+* As an admin I want to identify all *VENDOR* based images in the cluster:
+  * I create new `classification` using the *VENDOR* ImageStream(Tag)
+  * `ImageClassificationController` discovers all ancestors of this Image (or any previous
+     version of this Image) in the cluster
+  * `ImageClassificationController` applies label/annotations to the Image(s) that include
+    reference to *VENDOR*
+
+* As an admin I want to group the classifications together, so I can send "key=value" queries, for example:
+  * I want to get all images with `os=rhel7` or `vendor=redhat` or `lang=python`:
+  * For that I create a new classification group named “os” and provide an “selector” that matches the labels
+    in the target “classification” objects.
+
+
+### Proposed API Changes
+
+For the implementation, two new API types are needed:
+
+```go
+// ImageClass provides a single Image "class" a context that a
+// certain group if Images shares. For example "rhel7" or "python".
+// To create an ImageClass, you have to point it to an initial
+// or "target" image first. From that point, the ImageClassController
+// will track the history of that Image.
+type ImageClass struct{
+		unversioned.TypeMeta
+		kapi.ObjectMeta
+
+		Current      ImageChain   
+		History      []ImageChain 
+}
+
+// ImageClassGroup groups the various ImageClass objects into a
+// logical group. For example a "vendor" group groups all ImageClass
+// objects that matches the defined selector.
+type ImageClassGroup struct {
+		unversioned.TypeMeta
+		kapi.ObjectMeta
+
+		Selector: map[string]string
+}
+
+// ImageChain represents an internal state of the Image layers and
+// the chains based on the image layers.
+type ImageChain struct {
+		Chains    []string
+		LastChain string
+}
+```
+
+The `ImageClass` object is needed to keep the historical informations about the
+image persistent as the in-memory image chain index will only contain the
+images and chains that are currently available in the cluster.
```

Comment:
How is a chain in the `History` useful on its own if all the images referencing it are deleted?

Is it because a new association may be added in the future which will bring the chain back to life?